keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,93 @@
+
+It would be easier and expected to allow for changes to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+Relevant Issues:
+
+* https://github.com/kubernetes/kubernetes/issues/71477
+* https://github.com/kubernetes/kubernetes/issues/72198
+
+### Goals
+
+Allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by using an admission controller to check if the associated StorageClasses have `allowVolumeExpansion = true`.
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
+
+While updating a pod, the StatefulSet controller will update a referenced persistent volume claim object if its storage request in the associated `volumeClaimTemplate` has been increased.
+
+The Statefulset controller will delete and recreate the pod after the `FileSystemResizePending` condition is True on all updated persistent volume claims it references. However, if the `ExpandInUsePersistentVolumes` feature is enabled and the volume supports online expansion the controller will not wait for the `FileSystemResizePending` condition. 
+
+Note that the Statefulset controller does not support in place updates for any pod modifications. Therefore, even if all referenced volumes support online expansion the associated pods will still need to be deleted and recreated before completing the update. Support for in place updates may be introduced at a later point in time. Additionally, the best way too handle volume types that can only be expanded offline needs to be determined. These volume types require the pod to be offline before volume expansion can be initiated.
+
```

Comment:
So if I understand this correctly - stateful set controller will *ALWAYS* delete and re-recreate pods? If `ExpandInUsePersistentVolumes` feature is enabled it will not wait for `FileSystemResizePending` and it could delete and re-create the pods at any point in time? This sounds racy.  I know we have some checks in place to ensure we don't allow pod deletion while an operation is ongoing at kubelet but it is better to avoid it?

Reply:
Yes as @kow3ns mentioned above the workloads controllers don't to in place updates for any other container or Pod modifications today. More specifically the Statefulset controller will always delete and recreate the pod when update and current revision of the statefulset are different.  So until this functionality is built in general to allow for stable in-place updates of pods (and associated resources) an efficient workaround would be necessary for the issue you mentioned above. 

What we could do is when we recognize that the specific volume type supports online expansion and  `ExpandInUsePersistentVolumes` is enabled we would "wait" for expansion to complete before letting the controller delete the pod (just like we do [here](https://github.com/kubernetes/kubernetes/pull/71384/files#diff-988743e0eb9313bcfa8fd43ac3a0977fR235) for the `FileSystemResizePending` during offline expansion) . This won't break the current flow of when and how pods are deleted and recreated by the controller and would still avoid the issue of the pod being deleted during file system resize.

Reply:
Yeah I think so. It makes sense to wait for resizing to finish before restarting the pod. Lets update spec to reflect this.